### DISCUSSION TITLE
Custom Operator with TaskFlow

### DISCUSSION THREAD

**Discussion by zander1976** [apache/airflow#39333] [Q&A]
Good Afternoon,
I apologize is this has been covered before. I did search and review all the airflow code and astro code and no luck. I am sure it's something silly that I am doing. 

I created a simple operator:
```
class SPSSMOperator(BaseOperator):
    @apply_defaults
    def __init__(self, filename, parameters=None, *args, **kwargs):
        super(SPSSMOperator, self).__init__(*args, **kwargs)
        self.filename = filename
        self.parameters = parameters or {}

    def execute(self, context):
        """
        Execute the SPSSM task.
        """
        self.log.info("Executing SPSSM task with filename: %s", self.filename)
        self.log.info("Parameters: %s", self.parameters)
        spssm_result = f"Executed SPSSM task for {self.filename} with parameters: {self.parameters}"
        return spssm_result
```

and it seems to work fine with:
```
    spssm_task = SPSSMOperator(
        task_id="spss_task", 
        filename='file.cr', 
        parameters={'asdf': 'asdf', 'asdfasdf': 'asdfasdf'})
```

I wanted to register an @task.spssm like the following:
```
    @task.spssm(filename="file.cr", parameters={"asdf":"asdf", "asdfasdf":"asdfasdf"})
    def execute_spssm():
        ...
```

I then created the wrapper class and registration class:
```
class SPSSMDecoratedOperator(DecoratedOperator, SPSSMOperator):
    custom_operator_name: str = "@task.spssm"
    

def spssm_task(python_callable: Callable | None = None, **kwargs) -> TaskDecorator:
    return task_decorator_factory(
        python_callable=python_callable,
        decorated_operator_class=SPSSMDecoratedOperator,
        **kwargs,
    )

def get_provider_info():
    return {
        "package-name": "custom",
        "name": "SPSSM",
        "description": "SPSSM Executor",
        "task-decorators": [
            {
            "name": "spssm",
            "class-name": "spssm_task",
            }
        ],
        "versions": ["1.0.0"]
    }

```

All the code can be found in dags/custom/SPSSMOperator.py and I do have an empty __init__.py file in that directory.

I keep getting:
AttributeError: task decorator 'spssm' not found

Thank you so much for your help. SPSSM is a scripting language that runs to produce statistical data. 

Ben

**Comment by zander1976** [comment]
Ohh, I am using Alma linux 8.9 with Python3.9 and Airflow 2.8

**Reply by Taragolis** [reply]
You should:
1. Create your own provider
2. Register decorator within the own provider
3. Build python package (wheel or sdist)
4. Install it within the same environment as an Airflow (pip install ...)

Registration providers and component happen though [package entry points](https://packaging.python.org/en/latest/specifications/entry-points/)
